### PR TITLE
RavenDB-21344 Aligning how both Corax & Lucene handle full text searc…

### PR DESCRIPTION
**Pull request by ayende**
…h queries using wildcards

Better DebugView messages in the query plan

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21344 

### Additional description

We previously did nasty things to analyzed terms with wild cards
This worked, to an extent, as long as you didn't use a full text analyzer.

Now we apply the following logic for wild card / contains / prefix and ends with queries inside `search()`:

* If you are using `exact` analyzer, we'll use `exact`
* If you are using _any other_ analyzer, we'll use the `lower case` analyzer, and that is _it_. 

See also: https://ayende.com/blog/191841-B/understanding-query-processing-and-wildcards-in-ravendb

### CODE REVIEW

File: src/Corax/Analyzers/Analyzer.cs [hunk 1]
```diff
@@ -75,6 +76,8 @@ public static ArrayPool<Token> TokensPool
             float tokenBufferMultiplier = 1;
             foreach( var transformer in transformers)
             {
+                IsExactAnalyzer = transformer is ExactTransformer;
```
Comment (ppekrol) [inline src/Corax/Analyzers/Analyzer.cs]: this sets that according to last transformer. Is this correct?

Comment (ayende) [reply src/Corax/Analyzers/Analyzer.cs]: We only use 1 right now, but you are correct WRT future. Fixed.

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Common.cs [hunk 2]
```diff
@@ -37,6 +38,7 @@ private MultiTermMatch MultiTermMatchBuilder<TTermProvider>(in FieldMetadata fie
         {
             seekKey = _fieldsTree.Llt.AcquireCompactKey();
             seekKey.Set(seekTerm.AsReadOnlySpan());
+            seekKey.ChangeDictionary(terms.DictionaryId);
         }
```
Comment (redknightlois) [inline src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Common.cs:42]: Why do we need to change the dictionary or better phrased, if that is needed, why it didn't fail before?

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs [hunk 3]
```diff
@@ -408,7 +408,7 @@ private static IQueryMatch ToCoraxQuery(Parameters builderParameters, QueryExpre
                     var right = ToCoraxQuery(builderParameters, @where.Right, ref builderParameters.StreamingDisabled, exact);
 
                     builderParameters.BuildSteps?.Add(
-                        $"OR operator: left - {left.GetType().FullName} ({left}) assembly: {left.GetType().Assembly.FullName} assemby location: {left.GetType().Assembly.Location} , right - {right.GetType().FullName} ({right}) assemlby: {right.GetType().Assembly.FullName} assemby location: {right.GetType().Assembly.Location}");
+                        $"OR operator: left - {left.GetType().FullName} ({left}) assembly: {left.GetType().Assembly.FullName} assembly location: {left.GetType().Assembly.Location} , right - {right.GetType().FullName} ({right}) assemlbly: {right.GetType().Assembly.FullName} assembly location: {right.GetType().Assembly.Location}");
```
Comment (redknightlois) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs:412]: Typo on assembly.

**Comment by redknightlois** [comment]
LGTM

File: src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Common.cs (hunk 2 above)
Comment (ayende) [reply src/Corax/IndexSearcher/MultiTermMatches/IndexSearcher.MultiTermMatches.Common.cs:42]: This is so the `Inspect` call will tell us what is the actual key that is being used in terms.
That shouldn't cause any affect unless you are using query timings.

**Comment by arekpalinski** [comment]
test this please

**Review by arekpalinski — APPROVED** [review]

**Comment by arekpalinski** [comment]
Note that we change Lucene behavior as well. @ppekrol @maciejaszyk  please review.

I'm merging it for testing.

**Pull request merged**